File: 652.cpp
// Given a binary tree, return all duplicate subtrees. For each kind of
// duplicate subtrees, you only need to return the root node of any one of them.
//
// Two trees are duplicate if they have the same structure with same node
// values.
//
// Example 1:
//
//         1
//        / \
//       2   3
//      /   / \
//     4   2   4
//        /
//       4
// The following are two duplicate subtrees:
//
//       2
//      /
//     4
// and
//
//     4
// Therefore, you need to return above trees' root in the form of a list.

#include <iostream>
#include <map>
#include <vector>

using namespace std;

class Solution {
public:
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        map<string, vector<TreeNode*>> m;
        vector<TreeNode*> res;
        serialize(root, m, res);
        return res;
    }

    string serialize(TreeNode* root,
                     map<string, vector<TreeNode*>>& m,
                     vector<TreeNode*>& res) {
        if (!root) {
            return "";
        }

        string s = "[" + serialize(root->left, m, res) + to_string(root->val)
                   + serialize(root->right, m, res) + "]";
        if (m[s].size() == 1) {
            res.push_back(root);
        }
        m[s].push_back(root);
        return s;
    }
};

int main() {}

File: Python/652.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def findDuplicateSubtrees(self, root):
        """
        :type root: TreeNode
        :rtype: List[TreeNode]
        """
        res = []
        count = defaultdict(int)
        
        def postorder(root):
            if not root:
                return "#"
            id = str(root.val) + ","+ postorder(root.left) + "," + postorder(root.right)
            count[id] +=1
            if count[id] == 2:
                res.append(root)
            return id
        postorder(root)
        return res
